File: Java/src/session2/td/heritageMultiple/AlgebreNatRecursifZero.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public interface AlgebreNatRecursifZero extends Nat{

    default Nat somme(Nat x) {
        return x;
    }
    default Nat zero() {
        return this;
    }
    default Nat produit(Nat pNat) {
        return this.creerZero();
    }
    default Nat un() {
        return this.creerSuccesseur(this);
    }
    default Nat modulo(Nat pNat) {
        return this.creerZero();
    }
    default Nat div(Nat pNat) {
        return this.creerZero();
    }
    default boolean estEgal(Object o) {
        if (!(o instanceof Nat))
            return false;
        Nat x = (Nat) o;
        return x.estNul();
    }
    default String representer() {
        return "0";
    }
}

File: TypeScript/src/session2.td/heritageMultiple/AlgebreNatRecursifZero.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AlgebreNatRecursifZero_default = /** @class */ (function () {
    function AlgebreNatRecursifZero_default() {
    }
    AlgebreNatRecursifZero_default.prototype.zero = function () {
        return this.creerZero();
    };
    AlgebreNatRecursifZero_default.prototype.un = function () {
        return this.creerSuccesseur(this.creerZero());
    };
    AlgebreNatRecursifZero_default.prototype.somme = function (x) {
        return x;
    };
    AlgebreNatRecursifZero_default.prototype.produit = function (x) {
        return this.creerZero();
    };
    AlgebreNatRecursifZero_default.prototype.modulo = function (x) {
        return this.creerZero();
    };
    AlgebreNatRecursifZero_default.prototype.div = function (x) {
        return this.creerZero();
    };
    AlgebreNatRecursifZero_default.prototype.equals = function (o) {
        return this.val() == o.val();
    };
    return AlgebreNatRecursifZero_default;
}());
exports.AlgebreNatRecursifZero_default = AlgebreNatRecursifZero_default;
